homeassistant/components/goodwe/sensor.py

```diff
@@ -0,0 +1,257 @@
+        {vol.Required(ATTR_GRID_EXPORT_LIMIT): vol.Coerce(int)},
+        "set_grid_export_limit",
+    )
+
+    return True
+
+
+class InverterEntity(CoordinatorEntity, SensorEntity):
+    """Entity representing the inverter instance itself."""
+
+    _MAIN_ENTITY_SENSOR = "ppv"
+
+    def __init__(self, coordinator, device_info, inverter):
+        """Initialize the main inverter entity."""
+        super().__init__(coordinator)
+        self.entity_id = f".{DOMAIN}_inverter"
+        self._attr_unique_id = f"{DOMAIN}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+        self._attr_icon = "mdi:solar-power"
+        self._attr_name = "PV Inverter"
+        self._attr_native_unit_of_measurement = POWER_WATT
+        self._inverter = inverter
+
+    async def set_work_mode(self, work_mode: int):
+        """Set the inverter work mode."""
+        await self._inverter.set_work_mode(work_mode)
+
+    async def set_ongrid_battery_dod(self, ongrid_battery_dod: int):
+        """Set the on-grid battery dod."""
+        await self._inverter.set_ongrid_battery_dod(ongrid_battery_dod)
+
+    async def set_grid_export_limit(self, grid_export_limit: int):
+        """Set the grid export limit."""
+        await self._inverter.set_grid_export_limit(grid_export_limit)
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self.coordinator.data is not None
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        if self.coordinator.data is not None:
+            new_value = self.coordinator.data.get(self._MAIN_ENTITY_SENSOR)
+            # If no new value was provided, keep the previous
+            if new_value is not None:
+                self._attr_native_value = new_value
+
+        return self._attr_native_value
+
+    @property
+    def state_attributes(self):
+        """Return the inverter state attributes."""
+        return self.coordinator.data
+
+    @property
+    def extra_state_attributes(self):
+        """Return the inverter state attributes."""
+        data = {
+            "model": self._inverter.model_name,
+            "serial_number": self._inverter.serial_number,
+            "firmware_version": self._inverter.software_version,
+            "arm_version": self._inverter.arm_version,
+        }
+        return data
+
+
+class InverterSensor(CoordinatorEntity, SensorEntity):
+    """Entity representing individual inverter sensor."""
+
+    def __init__(self, coordinator, device_info, description, inverter):
+        """Initialize an inverter sensor."""
+        super().__init__(coordinator)
+        self.entity_description = description
+        self.entity_id = f".{DOMAIN}_{description.key}"
+        self._attr_unique_id = f"{DOMAIN}-{description.key}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+
+    @property
+    def available(self):
```

Comment:
This is consequence of the way of `async_update_data` is implemented. https://github.com/mletenay/core/blob/ceb22baff6917c341d9071106ea6e05859fb503c/homeassistant/components/goodwe/__init__.py#L72
When the inverter is not reachable, the `async_update_data` will not raise `UpdateFailed` as may be usual, but will pretend the update did not fail, but it returns None instead.
This was done to avoid excessive logging of those update failures, since for non-battery PV inverters, they are naturally hibernated for half of the day (night) and will resume responding back in the morning.

Reply:
The 3 tries make sense. But I am also confused about the reason not to raise UpdateFailed from the 4th attempt onwards.
Logging should still only occur just once in the evening and once in the morning.

Reply:
Or do you consider those two log entries to be too much?

Reply:
I see your point, I've looked into `DataUpdateCoordinator` and realized it logs only on first UpdateFailed and is silent until it responds back.
I will change it.

Reply:
I do not agree.
Even if it only generates 2 warnings a day, it is still annoying since the warnings are not true, I do not need to take any action, it is expected behaviour that the Inverter turns off at night.
Besides we will probably get a ton of users that have no experiance which will make issues regarding these warnings, while it is just expected behaviour.

Therefore I think it schould be a debug log at most, and I would catch the errors and use the available property as it is now.

Reply:
If you ho down that route, maybe create a custom coordinator class which inherits the base coordinator and do the override there rather than in sensor.py

Reply:
I made the change to standard behavior with 1 log entry per.
I never really liked the previous code distinguishing None and empty maps, it wasn't very readable.